File: proofload-core/src/main/kotlin/io/github/matthewjones372/proofload/Concurrency.kt
package io.github.matthewjones372.proofload

import kotlin.time.DurationUnit

/**
 * The users a run had in flight, beside the number its own throughput and
 * latency say it should have had.
 *
 * Little's law — L = λW — is arithmetic rather than a model: over a window
 * that starts and ends empty it holds for any arrival process and any service
 * distribution. So the two sides disagreeing is not a fact about the target,
 * it is a fact about the measurement, and every other number on the page is
 * suspect until it is explained.
 *
 * Absent carries the reason, as [Tail] and [Headroom] do. A run that cannot be
 * checked says so rather than reporting a ratio nobody can read.
 */
sealed interface Concurrency {

    data class Measured(
        /** Users running, averaged over the samples taken in the segment. */
        val observed: Double,
        /** Throughput times mean service time: what was outstanding on the wire. */
        val fromServiceTime: Double,
        /** The same against the clock that counts from the promised departure. */
        val fromResponseTime: Double,
        /** How many samples the observed side rests on. */
        val samples: Int,
    ) : Concurrency {

        /** What was measured over what was predicted: one where the law holds. */
        val ratio: Double get() = if (fromServiceTime == 0.0) 0.0 else observed / fromServiceTime

        /**
         * The queue the generator itself was holding, in requests.
         *
         * `responseTime = serviceTime + schedulingDelay` exactly, so the
         * difference between the two predictions is throughput times the mean
         * lateness — the backlog, counted the way the users are.
         */
        val backlog: Double get() = fromResponseTime - fromServiceTime

        /** Whether the two sides agree to within what this measurement can resolve. */
        val agrees: Boolean get() = ratio >= 1.0 - LAW_TOLERANCE && ratio <= 1.0 + LAW_TOLERANCE
    }

    data class Absent(val because: String) : Concurrency
}

/**
 * Little's law over the segment this run settled into, or why it cannot be
 * asked.
 *
 * Judged over the steady segment rather than the whole run: the identity is
 * exact over a window that starts and ends empty, and a run's opening seconds
 * are neither.
 */
val RunResult.concurrency: Concurrency
    get() {
        if (plan.pauses) {
            return Concurrency.Absent(
                "this scenario parks its users between steps, so a user in flight is not a request in flight",
            )
        }
        val settled = steady
        val window = settled.timeline.size.toDouble()
        if (window <= 0.0) return Concurrency.Absent("nothing was recorded second by second, so there is no window")

        val readings = usersInFlight.drop(usersInFlight.size - settled.usersInFlight.size).filterNotNull()
        if (readings.isEmpty()) return Concurrency.Absent("no user count was sampled, so there is nothing to check")

        val throughput = settled.count / window
        val service = settled.steps.values.map { it.serviceTime }.merged()
        val response = settled.steps.values.map { it.responseTime }.merged()
        if (service.count == 0L) return Concurrency.Absent("nothing was measured in the segment this run settled into")

        return Concurrency.Measured(
            observed = readings.average(),
            fromServiceTime = throughput * service.mean.toDouble(DurationUnit.SECONDS),
            fromResponseTime = throughput * response.mean.toDouble(DurationUnit.SECONDS),
            samples = readings.size,
        )
    }

/**
 * How far the two sides may differ and still be called agreement.
 *
 * The same figure [SteadyState.TOLERANCE] uses, and for the same reason: it is
 * what the coarse timeline can resolve. Symmetric, though the derived side is
 * biased high — every mean is read off bucket tops — because the measured side
 * is a one-hertz sample of a count that moves, and that error runs both ways.
 * A one-sided gate would fire on sampling noise and call it a bug in the tool.
 */
const val LAW_TOLERANCE: Double = SteadyState.TOLERANCE

File: proofload-scala/src/main/scala/io/github/matthewjones372/proofload/scala/Concurrency.scala
package io.github.matthewjones372.proofload.scala

import io.github.matthewjones372.proofload.ConcurrencyKt
import io.github.matthewjones372.proofload.RunResult
import io.github.matthewjones372.proofload.Concurrency as Core

/**
 * The users a run had in flight, beside the number its own throughput and
 * latency say it should have had.
 *
 * Little's law is arithmetic rather than a model, so the two sides disagreeing
 * is a fact about the measurement and not about the target: every other number
 * on the page is suspect until it is explained. A run the law cannot be asked
 * of says why, rather than reporting a ratio nobody can read.
 */
sealed trait Concurrency

object Concurrency:

  /** What the law found. It holds no number of its own: every one is core's. */
  final class Measured private[scala] (private val measured: Core.Measured) extends Concurrency:

    /** Users running, averaged over the samples taken in the segment. */
    def observed: Double = measured.getObserved

    /** Throughput times mean service time: what was outstanding on the wire. */
    def fromServiceTime: Double = measured.getFromServiceTime

    /** The same against the clock that counts from the promised departure. */
    def fromResponseTime: Double = measured.getFromResponseTime

    /** How many samples the observed side rests on. */
    def samples: Int = measured.getSamples

    /** What was measured over what was predicted: one where the law holds. */
    def ratio: Double = measured.getRatio

    /** The queue the generator itself was holding, in requests. */
    def backlog: Double = measured.getBacklog

    /** Whether the two sides agree to within what this measurement can resolve. */
    def agrees: Boolean = measured.getAgrees

    override def toString: String = measured.toString

  /** Why the law could not be asked of this run. */
  final case class Absent(because: String) extends Concurrency

extension (result: RunResult)

  /** Little's law over the segment this run settled into, or why it cannot be asked. */
  def concurrency: Concurrency = ConcurrencyKt.getConcurrency(result) match
    case measured: Core.Measured => Concurrency.Measured(measured)
    case absent: Core.Absent => Concurrency.Absent(absent.getBecause)
